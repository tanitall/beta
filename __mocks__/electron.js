module.exports = {
  app: {
    getPath: () => {
      return 'C:\\tmp\\mock_path';
    }
  },
};
