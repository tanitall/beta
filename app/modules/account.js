import { getAccountsFromWIFKey } from "neon-js";

// Constants
export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";
export const SET_DECRYPTING = "SET_DECRYPTING";
export const SET_KEYS = "SET_KEYS";

// Actions
export function login(wif) {
  return {
    type: LOGIN,
    wif: wif
  };
}

export function logout() {
  return {
    type: LOGOUT
  };
}

export function decrypting(bool) {
  return {
    type: SET_DECRYPTING,
    state: bool
  };
}

export function setKeys(keys) {
  return {
    type: SET_KEYS,
    keys
  };
}

// Reducer that manages account state (account now = private key)
export default (
  state = {
    wif: null,
    address: null,
    loggedIn: false,
    redirectUrl: null,
    decrypting: false,
    accountKeys: []
  },
  action
) => {
  switch (action.type) {
    case LOGIN:
      let loadAccount;
      try {
        loadAccount = getAccountsFromWIFKey(action.wif)[0];
      } catch (e) {
        loadAccount = -1;
      }
      if (
        loadAccount === -1 ||
        loadAccount === -2 ||
        loadAccount === undefined
      ) {
        return { ...state, wif: action.wif, loggedIn: false };
      }
      return {
        ...state,
        wif: action.wif,
        address: loadAccount.address,
        loggedIn: true,
        decrypting: false
      };
    case LOGOUT:
      return {
        ...state,
        wif: null,
        address: null,
        loggedIn: false,
        decrypting: false
      };
    case SET_DECRYPTING:
      return { ...state, decrypting: action.state };
    case SET_KEYS:
      return { ...state, accountKeys: action.keys };
    default:
      return state;
  }
};
