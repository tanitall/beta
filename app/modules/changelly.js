"use strict";
module.exports = (function() {
  let URL = "https://api.changelly.com";
  let io = require("socket.io-client");
  let jayson = require("jayson");
  let crypto = require("crypto");
  let client = jayson.client.https(URL);

  function Changelly(apiKey, apiSecret) {
    this._id = function() {
      return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
        c
      ) {
        let r = (Math.random() * 16) | 0,
          v = c == "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      });
    };

    this._sign = function(message) {
      return crypto
        .createHmac("sha512", apiSecret)
        .update(JSON.stringify(message))
        .digest("hex");
    };

    this._request = function(method, options, callback) {
      let id = this._id();
      let message = jayson.utils.request(method, options, id);
      client.options.headers = {
        "api-key": apiKey,
        sign: this._sign(message)
      };

      client.request(method, options, id, function(err, response) {
        callback(err, response);
      });
    };

    let self = this;

    this._socket = io.connect(URL, {
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      reconnectionAttempts: "Infinity"
    });

    this._socket.on("connect", function() {
      let message = {
        Login: {}
      };

      self._socket.emit("subscribe", {
        apiKey,
        sign: self._sign(message),
        message
      });
    });
  }

  Changelly.prototype = {
    getCurrencies(callback) {
      return this._request("getCurrencies", {}, callback);
    },
    createTransaction(from, to, address, amount, extraId, callback) {
      let params = {
        from,
        to,
        address,
        amount,
        extraId
      };

      return this._request("createTransaction", params, callback);
    },
    getMinAmount(from, to, callback) {
      let params = {
        from,
        to
      };

      return this._request("getMinAmount", params, callback);
    },
    getExchangeAmount(from, to, amount, callback) {
      let params = {
        from,
        to,
        amount
      };

      return this._request("getExchangeAmount", params, callback);
    },
    getTransactions(limit, offset, currency, address, extraId, callback) {
      let params = {
        limit,
        offset,
        currency,
        address,
        extraId
      };

      return this._request("getTransactions", params, callback);
    },
    getStatus(id, callback) {
      let params = {
        id
      };

      return this._request("getStatus", params, callback);
    },
    on(channel, callback) {
      this._socket.on(channel, callback);
    }
  };

  return Changelly;
})();
