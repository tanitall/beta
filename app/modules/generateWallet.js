import {
  getAccountsFromWIFKey,
  generatePrivateKey,
  getWIFFromPrivateKey,
  encrypt_wif
} from "neon-js";

// Constants
export const NEW_WALLET_KEYS = "NEW_WALLET_KEYS";
export const NEW_WALLET = "NEW_WALLET";
export const SET_GENERATING = "SET_GENERATING";
export const RESET_KEY = "RESET_KEY";

// Actions
export function newWalletKeys(passphrase) {
  return {
    type: NEW_WALLET_KEYS,
    passphrase
  };
}

export function newWallet(account) {
  return {
    type: NEW_WALLET,
    wif: account.wif,
    address: account.address,
    passphrase: account.passphrase,
    encryptedWif: account.encryptedWif
  };
}

export function generating(bool) {
  return {
    type: SET_GENERATING,
    state: bool
  };
}

export function resetKey() {
  return {
    type: RESET_KEY
  };
}

// Reducer used for state necessary to generating a wallet
export default (
  state = {
    wif: null,
    address: null,
    passphrase: null,
    encryptedWif: null,
    generating: false
  },
  action
) => {
  switch (action.type) {
    case NEW_WALLET_KEYS:
      const newPrivateKey = generatePrivateKey();
      const newWif = getWIFFromPrivateKey(newPrivateKey);
      const encryptedWif = encrypt_wif(newWif, action.passphrase);
      const loadAccount = getAccountsFromWIFKey(newWif);
      return {
        ...state,
        wif: newWif,
        address: loadAccount[0].address,
        passphrase: action.passphrase,
        encryptedWif
      };
    case NEW_WALLET:
      return {
        ...state,
        wif: action.wif,
        address: action.address,
        passphrase: action.passphrase,
        encryptedWif: action.encryptedWif,
        generating: false
      };
    case SET_GENERATING:
      return { ...state, generating: action.state };
    case RESET_KEY:
      return {
        wif: null,
        address: null,
        passphrase: null,
        encryptedWif: null,
        generating: false
      };
    default:
      return state;
  }
};
