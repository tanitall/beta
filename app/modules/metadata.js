// Constants
export const SET_HEIGHT = "SET_HEIGHT";
export const SET_NETWORK = "SET_NETWORK";
export const SET_EXPLORER = "SET_EXPLORER";

// Actions
export function setNetwork(net) {
  const network = net === "MainNet" ? "MainNet" : "TestNet";
  return {
    type: SET_NETWORK,
    net: network
  };
}

export function setBlockHeight(blockHeight) {
  return {
    type: SET_HEIGHT,
    blockHeight
  };
}

export function setBlockExplorer(blockExplorer) {
  return {
    type: SET_EXPLORER,
    blockExplorer
  };
}

// reducer for metadata associated with Neon
export default (
  state = { blockHeight: 0, network: "TestNet", blockExplorer: "Neotracker" },
  action
) => {
  switch (action.type) {
    case SET_HEIGHT:
      return { ...state, blockHeight: action.blockHeight };
    case SET_EXPLORER:
      console.log(action.blockExplorer);
      return { ...state, blockExplorer: action.blockExplorer };
    case SET_NETWORK:
      return { ...state, network: action.net };
    default:
      return state;
  }
};
