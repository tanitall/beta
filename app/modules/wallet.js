// Constants
export const SET_BALANCE = "SET_BALANCE";
export const SET_MARKET_PRICE = "SET_MARKET_PRICE";
export const RESET_PRICE = "RESET_PRICE";
export const SET_TRANSACTION_HISTORY = "SET_TRANSACTION_HISTORY";

// Actions
export function setBalance(neo, gas, price, combined, gasPrice) {
  return {
    type: SET_BALANCE,
    Neo: neo,
    Gas: gas,
    price: price,
    combined: combined,
    gasPrice: gasPrice
  };
}

export function setMarketPrice(price) {
  return {
    type: SET_MARKET_PRICE,
    price: price
  };
}

export function resetPrice() {
  return {
    type: RESET_PRICE
  };
}

export function setTransactionHistory(transactions) {
  return {
    type: SET_TRANSACTION_HISTORY,
    transactions
  };
}

// reducer for wallet account balance
export default (
  state = {
    Neo: 0,
    Gas: 0,
    transactions: [],
    price: "--",
    combined: "--",
    gasPrice: "--"
  },
  action
) => {
  switch (action.type) {
    case SET_BALANCE:
      return {
        ...state,
        Neo: action.Neo,
        Gas: action.Gas,
        price: action.price,
        combined: action.combined,
        gasPrice: action.gasPrice
      };
    case RESET_PRICE:
      return { ...state, price: "--" };
    case SET_MARKET_PRICE: //current market price action type
      let currentPrice;
      if (action.price !== undefined) {
        currentPrice = action.price;
      } else {
        currentPrice = "--";
      }
      return { ...state, price: currentPrice };
    case SET_TRANSACTION_HISTORY:
      return { ...state, transactions: action.transactions };
    default:
      return state;
  }
};
