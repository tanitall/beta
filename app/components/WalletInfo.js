import React, { Component } from "react";
import { connect } from "react-redux";
import Claim from "./Claim.js";
import MdSync from "react-icons/lib/md/sync";
import QRCode from "qrcode";
import Charts from "./Charts.js";
import { initiateGetBalance, intervals } from "../components/NetworkSwitch";
import { resetPrice } from "../modules/wallet";
import { sendEvent, clearTransactionEvent } from "../modules/transactions";
import { clipboard } from "electron";
import Copy from "react-icons/lib/md/content-copy";
import ReactTooltip from "react-tooltip";
import demoChart from "../images/demoChart.png";
import TopBar from "./TopBar";

// force sync with balance data
const refreshBalance = (dispatch, net, address) => {
  dispatch(sendEvent(true, "Refreshing..."));
  initiateGetBalance(dispatch, net, address).then(response => {
    dispatch(sendEvent(true, "Received latest blockchain information."));
    setTimeout(() => dispatch(clearTransactionEvent()), 1000);
  });
};

class WalletInfo extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount = () => {
    initiateGetBalance(this.props.dispatch, this.props.net, this.props.address);
    QRCode.toCanvas(this.canvas, this.props.address, { version: 5 }, err => {
      if (err) console.log(err);
    });
  };

  render = () => {
    if (this.props.address != null) {
      return (
        <div>
          <TopBar />
          <Charts />
          <div className="spacer" />
          {/* <div className="qrCode">
            <canvas id="qrCanvas" ref={node => (this.canvas = node)} />
          </div> */}
        </div>
      );
    } else {
      return null;
    }
  };
}

const mapStateToProps = state => ({
  neo: state.wallet.Neo,
  gas: state.wallet.Gas,
  address: state.account.address,
  net: state.metadata.network,
  price: state.wallet.price
});

WalletInfo = connect(mapStateToProps)(WalletInfo);

export default WalletInfo;
