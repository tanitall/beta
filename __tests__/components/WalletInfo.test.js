import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { SET_TRANSACTION_HISTORY, SET_BALANCE } from '../../app/modules/wallet';
import { SEND_TRANSACTION, CLEAR_TRANSACTION } from '../../app/modules/transactions';
import { SET_HEIGHT } from '../../app/modules/metadata';
import { SET_CLAIM } from '../../app/modules/claim';
import WalletInfo from '../../app/components/WalletInfo';

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { clipboard } from 'electron';
import { version } from '../../package.json';

const axiosMock = new MockAdapter(axios);
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version });
axiosMock
  .onGet('https://bittrex.com/api/v1.1/public/getticker?market=USDT-NEO')
  .reply(200, { result: { Last: 24.50 } });

jest.mock('electron', () => ({
  clipboard: {
    writeText: jest.fn(),
  },
}));
jest.mock('neon-js');

jest.unmock('qrcode');
import QRCode from 'qrcode';
QRCode.toCanvas = jest.fn();

const initialState = {
  account: {
    address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
  },
  metadata: {
    network: 'TestNet'
  },
  wallet: {
    Neo: 10,
    Gas: 1.0001,
    price: 25.48
  },
  claim: {
    claimAmount: 0.5
  }
};

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore()(state);

  let wrapper;
  if (shallowRender) {
    wrapper = shallow(<WalletInfo store={store} />);
  } else {
    wrapper = mount(
      <Provider store={store}>
        <WalletInfo />
      </Provider>
    );
  }

  return {
    store,
    wrapper
  };
};


describe('WalletInfo', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    done();
  });
  test('correctly renders data from state', (done) => {
    const { wrapper } = setup(initialState, false);

    const addressField = wrapper.find('.address');
    const fiatField = wrapper.find('.fiat');
    const neoField = wrapper.find('.amountNeo');
    const gasField = wrapper.find('.amountGas');

    expect(fiatField.text()).toEqual(`US ${initialState.wallet.price}`);
    expect(addressField.text().split('<')[0]).toEqual(initialState.account.address);
    expect(neoField.text()).toEqual(`${initialState.wallet.Neo}`);
    expect(gasField.text()).toEqual(`${Math.floor(initialState.wallet.Gas * 10000) / 10000}`);
    done();
  });
  test('copy to clipboard is getting called on click', (done) => {
    const { wrapper } = setup();
    const deepWrapper = wrapper.dive();

    expect(clipboard.writeText.mock.calls.length).toBe(0);
    deepWrapper.find('.copyKey').simulate('click');
    expect(clipboard.writeText.mock.calls.length).toBe(1);
    done();
  });
  test('refreshBalance is getting called on click', (done) => {
    const { wrapper, store } = setup();
    const deepWrapper = wrapper.dive();

    const actionTypes = [
      SEND_TRANSACTION,
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_BALANCE,
      CLEAR_TRANSACTION,
      SET_CLAIM
    ];
    deepWrapper.find('.refreshBalance').simulate('click');
    setTimeout(() => {
      const actions = store.getActions();
      expect(actions.length === 11).toEqual(true);
      actions.forEach(action => {
        expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
      });
      done();
    }, 1050)
  });
  test('calls the correct number of actions after mounting', (done) => {
    const { wrapper, store } = setup(initialState, false);
    const actionTypes = [
      SET_TRANSACTION_HISTORY,
      SET_HEIGHT,
      SET_BALANCE,
      SET_CLAIM
    ];
    setTimeout(() => {
      const actions = store.getActions();
      expect(actions.length).toEqual(4);
      actions.forEach(action => {
        expect(actionTypes.indexOf(action.type) > -1).toEqual(true)
      });
      done();
    }, 0);
  });
});
