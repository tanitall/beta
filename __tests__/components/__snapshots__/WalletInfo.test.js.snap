// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletInfo renders without crashing 1`] = `
<WalletInfo
  address="ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw"
  dispatch={[Function]}
  gas={1.0001}
  neo={10}
  net="TestNet"
  price={25.48}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
